### PR TITLE
ci: add auto release and changelog

### PR DESCRIPTION
**Pull request by dqhl76**
close: #1140 

tested on [fork repo](https://github.com/dqhl76/envd/releases/tag/v2.0.8)

### CODE REVIEW

**Comment by dqhl76** [comment]
@terrytangyuan PTAL

**Comment by gaocegege** [comment]
We can use `pip install --pre --upgrade envd` to install the new release.

**Comment by dqhl76** [comment]
> We can use `pip install --pre --upgrade envd` to install the new release.

Ok, I have changed hack/release-notes.md

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -192,3 +217,37 @@ jobs:
       env:
         BUILD_FUNC: ${{ matrix.build_func }}
         TAG_SUFFIX: ${{ matrix.TAG_SUFFIX }}
+  publish-release:
```
Comment (gaocegege) [inline .github/workflows/release.yml:220]: I think we already publish the bins. Do we need a new action for it?

File: .goreleaser.yaml [hunk 2]
```diff
@@ -111,6 +111,9 @@ docker_manifests:
   image_templates:
   - tensorchord/envd-from-scratch:v{{ .Version }}-amd64
   - tensorchord/envd-from-scratch:v{{ .Version }}-arm64v8
+release:
+  skip_upload: true
```
Comment (gaocegege) [inline .goreleaser.yaml:115]: I am wondering why we do not upload it with goreleaser.

Comment (dqhl76) [reply .goreleaser.yaml:115]: 1. This is because the format of the changelog uploaded using goreleaser is different from that of https://github.com/tensorchord/envd/blob/main/CHANGELOG.md.
2. @terrytangyuan said we need reuse this script: https://github.com/tensorchord/envd/blob/main/hack/changelog.sh https://github.com/tensorchord/envd/issues/1140#issuecomment-1300231365
3. We think we need put a redirect link in the release page
https://github.com/tensorchord/envd/issues/1140#issuecomment-1300754617

Comment (gaocegege) [reply .goreleaser.yaml:115]: Gotcha, thanks for your explanation!

**Review by gaocegege — COMMENTED** [review]
LGTM

/cc @terrytangyuan @VoVAllen @kemingy 

Thanks for your contribution! :tada: :+1:

**Comment by kemingy** [comment]
:cry: I feel it's too complex.

We already use the GitHub release note from our 1st release. I wonder if we can use the tools provided by GitHub.

For example, we can extract the latest 5 releases with:

```sh
gh release list --limit 5 | awk '{print $1}' | xargs -L 1 gh release view
```

```console
v0.2.5-alpha.2
gaocegege released this about 19 hours ago

  ## What's Changed

  • docs: updated CHANGELOG.md by @github-actions in
  https://github.com/tensorchord/envd/pull/1149
  • fix: horust cache by @kemingy in
  https://github.com/tensorchord/envd/pull/1151

  Full Changelog: https://github.com/tensorchord/envd/compare/v0.2.5-
  alpha.1...v0.2.5-alpha.2


Assets
checksums.txt                          1001 B
envd-sshd_0.2.5-alpha.2_Darwin_arm64   8.36 MiB
envd-sshd_0.2.5-alpha.2_Darwin_x86_64  8.44 MiB
envd-sshd_0.2.5-alpha.2_Linux_arm64    7.56 MiB
envd-sshd_0.2.5-alpha.2_Linux_i386     7.50 MiB
envd-sshd_0.2.5-alpha.2_Linux_x86_64   7.76 MiB
envd_0.2.5-alpha.2_Darwin_arm64        25.46 MiB
envd_0.2.5-alpha.2_Darwin_x86_64       25.62 MiB
envd_0.2.5-alpha.2_Linux_arm64         22.31 MiB
envd_0.2.5-alpha.2_Linux_i386          21.89 MiB
envd_0.2.5-alpha.2_Linux_x86_64        23.05 MiB

View on GitHub: https://github.com/tensorchord/envd/releases/tag/v0.2.5-alpha.2
......
```

Ref: https://cli.github.com/manual/gh_release_view

We can also get the JSON output. I think it's better than plain text.

**Comment by dqhl76** [comment]
> 😢 I feel it's too complex.
> 
> We already use the GitHub release note from our 1st release. I wonder if we can use the tools provided by GitHub.
> 
> For example, we can extract the latest 5 releases with:
> 
> ```shell
> gh release list --limit 5 | awk '{print $1}' | xargs -L 1 gh release view
> ```
> 
> ```
> v0.2.5-alpha.2
> gaocegege released this about 19 hours ago
> 
>   ## What's Changed
> 
>   • docs: updated CHANGELOG.md by @github-actions in
>   https://github.com/tensorchord/envd/pull/1149
>   • fix: horust cache by @kemingy in
>   https://github.com/tensorchord/envd/pull/1151
> 
>   Full Changelog: https://github.com/tensorchord/envd/compare/v0.2.5-
>   alpha.1...v0.2.5-alpha.2
> 
> 
> Assets
> checksums.txt                          1001 B
> envd-sshd_0.2.5-alpha.2_Darwin_arm64   8.36 MiB
> envd-sshd_0.2.5-alpha.2_Darwin_x86_64  8.44 MiB
> envd-sshd_0.2.5-alpha.2_Linux_arm64    7.56 MiB
> envd-sshd_0.2.5-alpha.2_Linux_i386     7.50 MiB
> envd-sshd_0.2.5-alpha.2_Linux_x86_64   7.76 MiB
> envd_0.2.5-alpha.2_Darwin_arm64        25.46 MiB
> envd_0.2.5-alpha.2_Darwin_x86_64       25.62 MiB
> envd_0.2.5-alpha.2_Linux_arm64         22.31 MiB
> envd_0.2.5-alpha.2_Linux_i386          21.89 MiB
> envd_0.2.5-alpha.2_Linux_x86_64        23.05 MiB
> 
> View on GitHub: https://github.com/tensorchord/envd/releases/tag/v0.2.5-alpha.2
> ......
> ```
> 
> Ref: https://cli.github.com/manual/gh_release_view
> 
> We can also get the JSON output. I think it's better than plain text.

Hi, I am happy to know this github command. However, I am a little confuse about where we should put this output? Replace [this](https://github.com/tensorchord/envd/blob/main/CHANGELOG.md)?

**Comment by kemingy** [comment]
> Hi, I am happy to know this github command. However, I am a little confuse about where we should put this output? Replace [this](https://github.com/tensorchord/envd/blob/main/CHANGELOG.md)?

Is it possible to run this command after each release?

`gh release list --limit 1 | awk '{print $1}' | gh release view >> CHANGELOG.md`

So we only need to guarantee the PR has the correct label (needs a GitHub Action). And we can still use the `release.yml` provided by GitHub.

WDYT cc @gaocegege

**Comment by dqhl76** [comment]
> > Hi, I am happy to know this github command. However, I am a little confuse about where we should put this output? Replace [this](https://github.com/tensorchord/envd/blob/main/CHANGELOG.md)?
> 
> Is it possible to run this command after each release?
> 
> `gh release list --limit 1 | awk '{print $1}' | gh release view >> CHANGELOG.md`
> 
> So we only need to guarantee the PR has the correct label (needs a GitHub Action). And we can still use the `release.yml` provided by GitHub.
> 
> WDYT cc @gaocegege

Actually, we can reopen [this pr](https://github.com/tensorchord/envd/pull/1155). :) and add this command and

**Comment by kemingy** [comment]
> Actually, we can reopen [this pr](https://github.com/tensorchord/envd/pull/1145) and add this command. :)

I think so. It should be easy to maintain. The current CI is already horrible enough.

WDYT @gaocegege

**Comment by terrytangyuan** [comment]
It will require manual labels on each PRs, which is redundant work since PRs already have good semantic titles for categorization. If the additional manual labeling is fine, then go ahead. Personally, I like a single-page changelog markdown file, so it's easy to look up without requiring other toolings.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -43,7 +43,32 @@ jobs:
...
+          mv dist/envd_linux_386/envd manifest/envd_${{ steps.get_version.outputs.VERSION }}_Linux_i386
+          mv dist/envd_darwin_amd64_v1/envd manifest/envd_${{ steps.get_version.outputs.VERSION }}_Darwin_x86_64
+          mv dist/envd_darwin_arm64/envd manifest/envd_${{ steps.get_version.outputs.VERSION }}_Darwin_arm64
+          mv dist/envd-sshd_linux_amd64_v1/envd-sshd manifest/envd-sshd_${{ steps.get_version.outputs.VERSION }}_Linux_x86_64
+          mv dist/envd-sshd_linux_arm64/envd-sshd manifest/envd-sshd_${{ steps.get_version.outputs.VERSION }}_Linux_arm64
+          mv dist/envd-sshd_linux_386/envd-sshd manifest/envd-sshd_${{ steps.get_version.outputs.VERSION }}_Linux_i386
+          mv dist/envd-sshd_darwin_amd64_v1/envd-sshd manifest/envd-sshd_${{ steps.get_version.outputs.VERSION }}_Darwin_x86_64
+          mv dist/envd-sshd_darwin_arm64/envd-sshd manifest/envd-sshd_${{ steps.get_version.outputs.VERSION }}_Darwin_arm64
+          mv dist/checksums.txt manifest/checksums.txt
+      - name: Upload publish manifest
```
Comment (terrytangyuan) [inline .github/workflows/release.yml:65]: Why is this needed? We can just use `dist/`

**Comment by kemingy** [comment]
> It will require manual labels on each PRs, which is redundant work since PRs already have good semantic titles for categorization. If the additional manual labeling is fine, then go ahead. 

There is an action in #1145 that provide the ability to label PR according to the title.

> Personally, I prefer less maintenance burden and single-page changelog markdown file, so it's easy to look up without requiring other toolings.

Agree. GitHub doesn't support searching the release note now. We can add the release note to the CHANGELOG.md file with the help of a GitHub Action.

File: .github/workflows/release.yml (hunk 3 above)
Comment (dqhl76) [reply .github/workflows/release.yml:65]: The file names in `dist/` have no version information. I've added these `mv` to keep the file name format consistent with the current release page.

**Comment by gaocegege** [comment]
> There is an action in #1145 that provides the ability to label PR according to the title.

LGTM, I also hope the release can be simple and clean.

We already encountered so many release issues.

**Comment by dqhl76** [comment]
I wrote a new changelog.sh and try to use `gh release view` to generate the CHANGELOG.md. https://github.com/dqhl76/envd/blob/main/new.md
https://gist.github.com/dqhl76/08ede5b4935e965c4f0ad2ad5c6b403d

However, compare with the [CHANGELOG.md](https://github.com/tensorchord/envd/blob/main/CHANGELOG.md) and changelog.sh we use currently, I don't find the new version is better enough. We just need a single markdown file to search. Maintaining a consistent format with the release page doesn't seem to be that important. But it is important to keep the internal formatting of this document consistent. The current version of CHANGELOG.md is very good in terms of consistency of internal formatting. So, I didn't change this part

If you think it is neccessary, I will replace it.

**Review by kemingy — COMMENTED** [review]
LGTM. May I know how to generate the release note when we release a new version? Do we need to run `goreleaser changelog`?

**Comment by dqhl76** [comment]
> LGTM. May I know how to generate the release note when we release a new version? Do we need to run `goreleaser changelog`?

same as before. Whenever the `release` workflow triggers.

**Comment by gaocegege** [comment]
Clean and elegant!

PTAL @terrytangyuan

**Review by gaocegege — APPROVED** [review]

**Review by terrytangyuan — APPROVED** [review]
Looks great! Thanks for your contributions!

**Comment by gaocegege** [comment]
I made a new release. But the changelog does not work as expected, I think. 

https://github.com/tensorchord/envd/releases/tag/v0.2.5-alpha.3

**Comment by dqhl76** [comment]
> I made a new release. But the changelog does not work as expected, I think.
> 
> https://github.com/tensorchord/envd/releases/tag/v0.2.5-alpha.3

I am not sure why it is not work. It was sussesful when I test in the fork repo.
I will check later.

**Comment by gaocegege** [comment]
Thanks!

**Comment by dqhl76** [comment]
https://discord.com/channels/974584200327991326/985401618277228604/1041629245337767956

**Comment by gaocegege** [comment]
@all-contributors Please add @dqhl76 for the code contribution
